app/src/main/java/protect/card_locker/PkpassParser.kt

```diff
@@ -0,0 +1,404 @@
+    private val balance: BigDecimal = BigDecimal(0)
+    private val balanceType: Currency? = null
+    private var cardId: String? = null
+    private var barcodeId: String? = null
+    private var barcodeType: CatimaBarcode? = null
+    private var headerColor: Int? = null
+    private val starStatus = 0
+    private val lastUsed: Long = 0
+    private val zoomLevel = DBHelper.DEFAULT_ZOOM_LEVEL
+    private var archiveStatus = 0
+
+    private var image: Bitmap? = null
+    private var logoSize = 0
+
+    init {
+        if (passContent != null) {
+            throw IllegalStateException("Pkpass instance already initialized!")
+        }
+
+        mContext = context
+
+        Log.i(TAG, "Received Pkpass file")
+        if (uri == null) {
+            Log.e(TAG, "Uri did not contain any data")
+            throw IOException(context.getString(R.string.errorReadingFile))
+        }
+
+        val inputStream = mContext.contentResolver.openInputStream(uri) ?: throw IOException(mContext.getString(R.string.errorReadingFile))
+        val zipInputStream = ZipInputStream(inputStream)
+
+        try {
+            var localFileHeader: LocalFileHeader
+            while ((zipInputStream.nextEntry.also { localFileHeader = it }) != null) {
+                // Ignore directories
+                if (localFileHeader.isDirectory) continue
+
+                // We assume there are three options, as per spec:
+                // language.lproj/pass.strings
+                // file.extension
+                // More directories are ignored
+                val filenameParts = localFileHeader.fileName.split('/')
+                if (filenameParts.size > 2) {
+                    continue
+                } else if (filenameParts.size == 2) {
+                    // Doesn't seem like a language directory, ignore
+                    if (!filenameParts[0].endsWith(".lproj")) continue
+
+                    val locale = filenameParts[0].removeSuffix(".lproj")
+
+                    translations[locale] = parseLanguageStrings(ZipUtils.read(zipInputStream))
+                }
+
+                // Not a language, parse as normal files
+                when (localFileHeader.fileName) {
+                    "logo.png" -> loadImageIfBiggerSize(1, zipInputStream)
+                    "logo@2x.png" -> loadImageIfBiggerSize(2, zipInputStream)
+                    "logo@3x.png" -> loadImageIfBiggerSize(3, zipInputStream)
+                    "pass.json" -> passContent = ZipUtils.readJSON(zipInputStream) // Parse this last, so we're sure we have all language info
+                }
+            }
+
+            checkNotNull(passContent) { "File lacks pass.json" }
+        } catch (e: Exception) {
+            throw e
+         } finally {
+            try {
+                inputStream.close()
+            } catch (e: IOException) {
+                Log.e(TAG, "Error closing inputStream", e)
+            }
+
+            try {
+                zipInputStream.close()
```

Comment:
Tip: With streams, you can use the kotlin's `.use` method to close them when done.
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html

Reply:
Example: https://gitlab.com/AuroraOSS/AuroraStore/-/blob/master/app/src/main/java/com/aurora/store/data/work/DownloadWorker.kt#L294

Reply:
Thanks, also for the pointer! It's basically Python's `with` statement, it does feel cleaner and helps scope more explicitly :)